dev-tools/omdb/src/bin/omdb/mgs/sensors.rs

```diff
@@ -0,0 +1,954 @@
+        }
+    }
+
+    fn format(&self, value: f32, parseable: bool) -> String {
+        if parseable {
+            format!("{value}")
+        } else {
+            match self.kind {
+                MeasurementKind::Speed => {
+                    format!("{value:0} RPM")
+                }
+                _ => {
+                    format!("{value:.2}{}", self.units())
+                }
+            }
+        }
+    }
+
+    fn to_kind_string(&self) -> &str {
+        match self.kind {
+            MeasurementKind::Temperature => "temp",
+            MeasurementKind::Power => "power",
+            MeasurementKind::Current => "current",
+            MeasurementKind::Voltage => "voltage",
+            MeasurementKind::InputCurrent => "input-current",
+            MeasurementKind::InputVoltage => "input-voltage",
+            MeasurementKind::Speed => "speed",
+        }
+    }
+
+    fn from_string(name: &str, kind: &str) -> Option<Self> {
+        let k = match kind {
+            "temp" | "temperature" => Some(MeasurementKind::Temperature),
+            "power" => Some(MeasurementKind::Power),
+            "current" => Some(MeasurementKind::Current),
+            "voltage" => Some(MeasurementKind::Voltage),
+            "input-current" => Some(MeasurementKind::InputCurrent),
+            "input-voltage" => Some(MeasurementKind::InputVoltage),
+            "speed" => Some(MeasurementKind::Speed),
+            _ => None,
+        };
+
+        if let Some(kind) = k {
+            Some(Sensor { name: name.to_string(), kind })
+        } else {
+            None
+        }
+    }
+}
+
+pub(crate) enum SensorInput<R> {
+    MgsClient(gateway_client::Client),
+    CsvReader(csv::Reader<R>, csv::Position),
+}
+
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SensorId(u32);
+
+#[derive(Debug)]
+pub(crate) struct SensorMetadata {
+    pub sensors_by_sensor: MultiMap<Sensor, SensorId>,
+    pub sensors_by_sensor_and_sp:
+        HashMap<Sensor, HashMap<SpIdentifier, SensorId>>,
+    pub sensors_by_id:
+        HashMap<SensorId, (SpIdentifier, Sensor, DeviceIdentifier)>,
+    pub sensors_by_sp: MultiMap<SpIdentifier, SensorId>,
+    pub work_by_sp:
+        HashMap<SpIdentifier, Vec<(DeviceIdentifier, Vec<SensorId>)>>,
+    #[allow(dead_code)]
+    pub start_time: Option<u64>,
+    pub end_time: Option<u64>,
+}
+
+struct SensorSpInfo {
+    info: Vec<(SpIdentifier, SpInfo)>,
+    time: u64,
+    latencies: Option<HashMap<SpIdentifier, Duration>>,
+}
+
+pub(crate) struct SensorValues {
+    pub values: HashMap<SensorId, Option<f32>>,
+    pub latencies: Option<HashMap<SpIdentifier, Duration>>,
+    pub time: u64,
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub(crate) enum DeviceIdentifier {
+    Field(usize),
+    Device(String),
+}
+
+impl DeviceIdentifier {
+    fn device(&self) -> &String {
+        match self {
+            Self::Device(ref device) => device,
+            _ => panic!(),
```

Comment:
This seems a little fishy (likewise the panic in `field()`). I'm tempted to propose one of two changes:

* Make this return `Option<&String>` (or `Option<&str>` since `&String` is kinda weird), and let the caller unwrap if they know somehow that this must be a device.
* (Maybe? This might be really annoying) Drop this enum entirely? If its users always know exactly which case they have (which they have to to use these methods), could they use the underlying types instead?

Reply:
So, I'm not as in to either of those. To me, this makes the consuming code more complicated -- and the `unwrap` to me would introduce clutter without meaningfully changing anything. Inclined to add a comment here rather than restructure or drop it...

Reply:
Sure; this isn't a hill I'm dying on, since this isn't a public API and misuse is pretty unlikely.

Reply:
I added a comment to clarify.